### PR TITLE
Speed up computing proof using for Admitted when trivial

### LINKED ISSUE #15858: It should not take 200 seconds to run `Admitted`
#### Description of the problem
In https://github.com/mit-plv/fiat-crypto/pull/1173/files#diff-0a7841db22fe2f52e8b832d49d801d1fb5fd44ef4e95ced2e867375da8599dffR3162 at line 3162 of src/Assembly/WithBedrock/Proofs.v, there's an `Admitted` that took over 200 seconds to run on Coq 8.14.0.  What is Coq doing?

(I'll update this with a permalink once the relevant PR in fiat-crypto is merged)
EDIT by skyskimmer: no need to wait https://github.com/mit-plv/fiat-crypto/blob/d35a2a87386be80a9b4095a61766d8524e35f3ab/src/Assembly/WithBedrock/Proofs.v#L3162

cc @ppedrot ?

#### Coq Version

8.14.0

### PR DESCRIPTION
**Pull request by SkySkimmer**
- when not in section, there is never anything to do.
- when in section but Unset Keep Admitted Variables, also skip evar
- normalizing the proof.

Fix #15858 ?

### CODE REVIEW

**Comment by SkySkimmer** [comment]
cc @JasonGross

**Comment by SkySkimmer** [comment]
@coqbot bench

File: vernac/declare.ml [hunk 1]
```diff
@@ -1990,18 +1990,20 @@ let get_keep_admitted_vars =
...
+    | None ->
+      match Proof.partial_proof iproof with
+      | pproof :: _ ->
+        let env = Global.env () in
+        let ids_typ = Environ.global_vars_set env typ in
+        (* [pproof] is evar-normalized by [partial_proof]. We don't
+           count variables appearing only in the type of evars. *)
+        let ids_def = Environ.global_vars_set env (EConstr.Unsafe.to_constr pproof) in
+        Some (Environ.really_needed env (Id.Set.union ids_typ ids_def))
+      | [] -> None
```
Comment (Alizter) [inline vernac/declare.ml:2006]:
```suggestion
let compute_proof_using_for_admitted proof typ iproof =
  if get_keep_admitted_vars () && Global.sections_are_opened () then
    match get_used_variables proof with
    | Some _ as x -> x
    | None ->
      match Proof.partial_proof iproof with
      | pproof :: _ ->
        let env = Global.env () in
        let ids_typ = Environ.global_vars_set env typ in
        (* [pproof] is evar-normalized by [partial_proof]. We don't
           count variables appearing only in the type of evars. *)
        let ids_def = Environ.global_vars_set env (EConstr.Unsafe.to_constr pproof) in
        Some (Environ.really_needed env (Id.Set.union ids_typ ids_def))
      | [] -> None
  else None
```

Comment (Alizter) [reply vernac/declare.ml:2006]: The negation and if made it a bit confusing to read. Hopefully this is clearer.

Comment (SkySkimmer) [reply vernac/declare.ml:2006]: I prefer having the short case first

**Comment by JasonGross** [comment]
I've merged https://github.com/mit-plv/fiat-crypto/pull/1173, so if you re-bench coq-fiat-crypto-with-bedrock, you should see the results in the bench (specifically in file src/Assembly/WithBedrock/Proofs.v)

**Comment by SkySkimmer** [comment]
https://gitlab.com/coq/coq/-/jobs/2256089077

**Comment by SkySkimmer** [comment]
It failed 
~~~
- /usr/bin/make --no-print-directory -C rupicola install
- make[1]: *** No rule to make target 'install'.  Stop.
- make: *** [Makefile:522: install-rupicola] Error 2
~~~

**Comment by JasonGross** [comment]
I've merged https://github.com/mit-plv/rupicola/pull/52, let's try again. @coqbot bench

**Comment by JasonGross** [comment]
Er, oops that probably benches everything, I'm not sure how to start the bench with just one package

**Comment by SkySkimmer** [comment]
No idea why but it seems it only started the bench with the right package

**Review by ejgallego — APPROVED** [review]

**Comment by ejgallego** [comment]
@SkySkimmer you let me know when you want this merged?

**Comment by SkySkimmer** [comment]
@coqbot bench

**Comment by SkySkimmer** [comment]
~~~
# /usr/bin/make --no-print-directory -C rupicola install
# Generating Makefile.coq
# Warning: No common logical root.
# Warning: In this case the -docroot option should be given.
# Warning: Otherwise the install-doc target is going to install files
# Warning: in orphan_coqutil_bedrock2_Rupicola
# make[1]: *** No rule to make target 'install_coqutil', needed by 'install'.  Stop.
~~~

**Comment by JasonGross** [comment]
Should be fixed by https://github.com/mit-plv/fiat-crypto/commits/8ab610ab293918c46c6e23c4be19690b957f392c / https://github.com/mit-plv/rupicola/commit/6fed25e3eef452ddfbc3c1107cb18925d58c5599, let's try yet again @coqbot bench

**Comment by JasonGross** [comment]
Failed again at https://gitlab.com/coq/coq/-/jobs/2267371305.  What's the error (and how do I find the error)?

**Comment by Alizter** [comment]
In the artifacts for the job, there is a logs folder which contains the stderr from the job:
```
[ERROR] The installation of coq-fiat-crypto-with-bedrock failed at "make EXTERNAL_REWRITER=1 EXTERNAL_COQPRIME=1 BINDIR=/home/gitlab-runner/builds/LBs8AQPP/0/coq/coq/_bench/opam.NEW/ocaml-base-compiler.4.07.1/bin install install-standalone-ocaml".

#=== ERROR while installing coq-fiat-crypto-with-bedrock.dev ==================#
# context              2.0.3 | linux/x86_64 | ocaml-base-compiler.4.07.1 | file:///home/gitlab-runner/builds/LBs8AQPP/0/coq/coq/_bench/new_coq_opam_archive/extra-dev
# path                 ~/builds/LBs8AQPP/0/coq/coq/_bench/opam.NEW/ocaml-base-compiler.4.07.1/.opam-switch/build/coq-fiat-crypto-with-bedrock.dev
# command              /usr/bin/make EXTERNAL_REWRITER=1 EXTERNAL_COQPRIME=1 BINDIR=/home/gitlab-runner/builds/LBs8AQPP/0/coq/coq/_bench/opam.NEW/ocaml-base-compiler.4.07.1/bin install install-standalone-ocaml
# exit-code            2
# env-file             ~/builds/LBs8AQPP/0/coq/coq/_bench/opam.NEW/log/coq-fiat-crypto-with-bedrock-19543-b76d46.env
# output-file          ~/builds/LBs8AQPP/0/coq/coq/_bench/opam.NEW/log/coq-fiat-crypto-with-bedrock-19543-b76d46.out
### output ###
# [...]
# INSTALL /home/gitlab-runner/builds/LBs8AQPP/0/coq/coq/_bench/opam.NEW/ocaml-base-compiler.4.07.1/.opam-switch/build/coq-fiat-crypto-with-bedrock.dev/rupicola/bedrock2/deps/coqutil/src/coqutil/Datatypes/Inhabited.glob /home/gitlab-runner/builds/LBs8AQPP/0/coq/coq/_bench/opam.NEW/ocaml-base-compiler.4.07.1/lib/coq//user-contrib/coqutil//Datatypes
# INSTALL /home/gitlab-runner/builds/LBs8AQPP/0/coq/coq/_bench/opam.NEW/ocaml-base-compiler.4.07.1/.opam-switch/build/coq-fiat-crypto-with-bedrock.dev/rupicola/bedrock2/deps/coqutil/src/coqutil/Datatypes/List.glob /home/gitlab-runner/builds/LBs8AQPP/0/coq/coq/_bench/opam.NEW/ocaml-base-compiler.4.07.1/lib/coq//user-contrib/coqutil//Datatypes
# INSTALL /home/gitlab-runner/builds/LBs8AQPP/0/coq/coq/_bench/opam.NEW/ocaml-base-compiler.4.07.1/.opam-switch/build/coq-fiat-crypto-with-bedrock.dev/rupicola/bedrock2/deps/coqutil/src/coqutil/Datatypes/Prod.glob /home/gitlab-runner/builds/LBs8AQPP/0/coq/coq/_bench/opam.NEW/ocaml-base-compiler.4.07.1/lib/coq//user-contrib/coqutil//Datatypes
# INSTALL /home/gitlab-runner/builds/LBs8AQPP/0/coq/coq/_bench/opam.NEW/ocaml-base-compiler.4.07.1/.opam-switch/build/coq-fiat-crypto-with-bedrock.dev/rupicola/bedrock2/deps/coqutil/src/coqutil/Datatypes/ListSet.glob /home/gitlab-runner/builds/LBs8AQPP/0/coq/coq/_bench/opam.NEW/ocaml-base-compiler.4.07.1/lib/coq//user-contrib/coqutil//Datatypes
# INSTALL /home/gitlab-runner/builds/LBs8AQPP/0/coq/coq/_bench/opam.NEW/ocaml-base-compiler.4.07.1/.opam-switch/build/coq-fiat-crypto-with-bedrock.dev/rupicola/bedrock2/deps/coqutil/src/coqutil/Byte.glob /home/gitlab-runner/builds/LBs8AQPP/0/coq/coq/_bench/opam.NEW/ocaml-base-compiler.4.07.1/lib/coq//user-contrib/coqutil/
# INSTALL /home/gitlab-runner/builds/LBs8AQPP/0/coq/coq/_bench/opam.NEW/ocaml-base-compiler.4.07.1/.opam-switch/build/coq-fiat-crypto-with-bedrock.dev/rupicola/bedrock2/deps/coqutil/src/coqutil/sanity.glob /home/gitlab-runner/builds/LBs8AQPP/0/coq/coq/_bench/opam.NEW/ocaml-base-compiler.4.07.1/lib/coq//user-contrib/coqutil/
# # findlib needs the package to not be installed, so we remove it before
# # installing it
# /usr/bin/make --no-print-directory -C bedrock2/bedrock2 install_bedrock2
# make[2]: *** No rule to make target 'install_bedrock2'.  Stop.
# make[1]: *** [Makefile:97: install_bedrock2] Error 2
# make: *** [Makefile:525: install-rupicola] Error 2

'opam install -v -b -j1 coq-fiat-crypto-with-bedrock' failed.
```

**Comment by JasonGross** [comment]
I *think* this means that everything built but did not install.  I've fixed this missing target, hopefully this time it'll actually work.  @coqbot bench

**Comment by SkySkimmer** [comment]
~~~
- /usr/bin/make -f Makefile.coq install
- COQDEP VFILES
- /home/gitlab-runner/builds/av_eR6za/0/coq/coq/_bench/opam.NEW/ocaml-base-compiler.4.07.1/.opam-switch/build/coq-fiat-crypto-with-bedrock.dev/rupicola/src/Rupicola/Examples/Arithmetic.vo does not exist
~~~
and many more "xxx does not exist"
(you need to look at the stdout file as the stderr is redacted)

**Comment by JasonGross** [comment]
This issue was https://github.com/coq/opam-coq-archive/pull/2137#issuecomment-1082620562, https://github.com/mit-plv/fiat-crypto/pull/1180 has been merged, let's try again @coqbot bench

**Comment by JasonGross** [comment]
It finally worked!
> 48.5390    0.0000  -48.5390  -100.00%  3154  coq-fiat-crypto-with-bedrock/src/Assembly/WithBedrock/Proofs.v.html

Looks like this PR is a winner

**Comment by ppedrot** [comment]
Out of curiosity, it's the call to `partial_proof` that takes so long?

**Comment by SkySkimmer** [comment]
It could also be partial_vars_set.

**Comment by ejgallego** [comment]
@coqbot: merge now

**Pull request merged**